emacs/config.org
#+title: My Emacs Configuration
#+author: Nick Salesky
#+PROPERTY: header-args:emacs-lisp :tangle init.el
#+STARTUP: overview

* About
This is my personal Emacs config that I've built with the "help" :) of many other open-source configs, blog posts, articles, etc. When possible, I try to attribute specific sections to the original source for both my benefit and anyone else's. I started using Emacs with the help of [[https://github.com/doomemacs/doomemacs][Doom Emacs]], so many of my keybindings and other settings are inspired by that.

One important thing is that I haven't really done any effort to shorten the Emacs loading time, so this config takes a while to start up (about 8 seconds or so on my machine). I usually just keep Emacs open or use a daemon so this isn't an issue for me, but maybe something for me to look into in the future!

Some specific configs I looked at include:

- [[https://github.com/dakra/dmacs][Dmacs]] (the structure of this config file is heavily inspired by this!)
- [[https://www.youtube.com/watch?v=SCPoF1PTZpI&t=896s][System Crafters]]

* Installation

Installation should be fairly straightforward. The first time Emacs runs this config, it should hopefully install all the necessary packages automatically. If it doesn't prompt you, you'll need to run ~M-x all-the-icons-install-fonts~.

* Emacs Initialization
Add my custom ELisp files to the load-path.
#+BEGIN_SRC emacs-lisp
(add-to-list 'load-path (expand-file-name "lisp" user-emacs-directory))
#+END_SRC 

Enable native compilation for all Elisp code
#+BEGIN_SRC emacs-lisp
(setq native-comp-deferred-compilation t)
#+END_SRC

Disable certain byte compiler warnings to cut down on the noise for (usually unnecessary) errors.

#+BEGIN_SRC emacs-lisp
(setq byte-compile-warnings '(not free-vars unresolved noruntime lexical make-local)
      native-comp-async-report-warnings-errors nil)
#+END_SRC 

Increase the garbage collection threshold and process data chunking for performance.

#+BEGIN_SRC emacs-lisp
(setq gc-cons-threshold 100000000
      read-process-output-max (* 1024 1024))
#+END_SRC 

** Set some basic UI settings earl

Disable UI elements
#+BEGIN_SRC emacs-lisp
(tool-bar-mode -1)
(menu-bar-mode -1)
(scroll-bar-mode -1)
(tooltip-mode -1)
(set-fringe-mode 10)
#+END_SRC

Disable the startup message
#+BEGIN_SRC emacs-lisp
(setq inhibit-startup-message t)
#+END_SRC 

Use short answers for prompts
#+BEGIN_SRC emacs-lisp
(setq use-short-answers t)
#+END_SRC 

Don't confirm nonexistent files or buffers
#+BEGIN_SRC emacs-lisp
(setq confirm-nonexistent-file-or-buffer nil)
#+END_SRC 

** custom.el
Create a ~custom.el~ file to store customization info instead of init.el, since this is a literal config.

#+BEGIN_SRC emacs-lisp
(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
#+END_SRC

* Package System Setup
I'm using =elpaca= with =use-package= to install and configure packages

Set up =elpaca=
#+BEGIN_SRC emacs-lisp
(defvar elpaca-installer-version 0.6)
(defvar elpaca-directory (expand-file-name "elpaca/" user-emacs-directory))
(defvar elpaca-builds-directory (expand-file-name "builds/" elpaca-directory))
(defvar elpaca-repos-directory (expand-file-name "repos/" elpaca-directory))
(defvar elpaca-order '(elpaca :repo "https://github.com/progfolio/elpaca.git"
                              :ref nil
                              :files (:defaults "elpaca-test.el" (:exclude "extensions"))
                              :build (:not elpaca--activate-package)))
(let* ((repo  (expand-file-name "elpaca/" elpaca-repos-directory))
       (build (expand-file-name "elpaca/" elpaca-builds-directory))
       (order (cdr elpaca-order))
       (default-directory repo))
  (add-to-list 'load-path (if (file-exists-p build) build repo))
  (unless (file-exists-p repo)
    (make-directory repo t)
    (when (< emacs-major-version 28) (require 'subr-x))
    (condition-case-unless-debug err
        (if-let ((buffer (pop-to-buffer-same-window "*elpaca-bootstrap*"))
                 ((zerop (call-process "git" nil buffer t "clone"
                                       (plist-get order :repo) repo)))
                 ((zerop (call-process "git" nil buffer t "checkout"
                                       (or (plist-get order :ref) "--"))))
                 (emacs (concat invocation-directory invocation-name))
                 ((zerop (call-process emacs nil buffer nil "-Q" "-L" "." "--batch"
                                       "--eval" "(byte-recompile-directory \".\" 0 'force)")))
                 ((require 'elpaca))
                 ((elpaca-generate-autoloads "elpaca" repo)))
            (progn (message "%s" (buffer-string)) (kill-buffer buffer))
          (error "%s" (with-current-buffer buffer (buffer-string))))
      ((error) (warn "%s" err) (delete-directory repo 'recursive))))
  (unless (require 'elpaca-autoloads nil t)
    (require 'elpaca)
    (elpaca-generate-autoloads "elpaca" repo)
    (load "./elpaca-autoloads")))
(add-hook 'after-init-hook #'elpaca-process-queues)
(elpaca `(,@elpaca-order))
#+END_SRC

Disable =package.el=
#+BEGIN_SRC emacs-lisp :tangle early-init.el
(setq package-enable-at-startup nil)
#+END_SRC

Install =use-package= support
#+BEGIN_SRC emacs-lisp
(elpaca elpaca-use-package
        ;; Enable :elpaca use-package keyword
        (elpaca-use-package-mode)
        ;; Assume :elpaca t unless otherwise specified
        (setq elpaca-use-package-by-default t))

;; Upgrade built-in packages
(elpaca transient)

;; Block until current queue processed
(elpaca-wait)
#+END_SRC


Set up environment variables
#+BEGIN_SRC emacs-lisp
(use-package exec-path-from-shell
  :config
  (exec-path-from-shell-initialize))
#+END_SRC 

* Personal Information
Set up the basic user information variables.

#+BEGIN_SRC emacs-lisp
(setq user-full-name "Nick Salesky"
      user-mail-address "nicksalesky@gmail.com")
#+END_SRC 

* Keybindings and Basic Feature Configuration
Here's where I configure some basic Emacs features with keybindings.

** Set up modifiers for Mac OS X
#+BEGIN_SRC emacs-lisp
(when (eq system-type 'darwin)
  (setq mac-option-key-is-meta t
        mac-command-key-is-meta t
        mac-command-modifier 'meta
        mac-option-modifier 'meta))
#+END_SRC

** Enable all disabled keybindings
#+BEGIN_SRC emacs-lisp
(setq disabled-command-function nil)
#+END_SRC

** Evil 😈

#+BEGIN_SRC emacs-lisp :tangle nil
(use-package evil
  ;; :ensure t
  :init
  (setq evil-want-keybinding nil
        evil-want-integration t
        evil-want-C-u-scroll t
        evil-want-C-d-scroll t
        evil-undo-system 'undo-redo
  )

  :config
  (evil-mode 1))

(use-package evil-collection
  :ensure t
  :after evil
  :config
  (evil-collection-init))

(use-package evil-surround
  :config
  (global-evil-surround-mode 1))

(elpaca-wait)
#+END_SRC

** General

#+BEGIN_SRC emacs-lisp :tangle nil
(use-package general
  :ensure t
  :config
  (general-evil-setup t)
  (general-create-definer ns/leader-def
    :keymaps '(normal visual emacs)
    :prefix "SPC"
    :global-prefix "C-SPC")
  (general-create-definer ns/local-leader-def
    :keymaps '(normal visual emacs)
    :prefix "SPC m"
    :global-prefix "C-SPC m"))
(elpaca-wait)
#+END_SRC

** Electric-pair-mode
Disable ~electric-pair-mode~ for a select subset of modes. 

#+BEGIN_SRC emacs-lisp
(electric-pair-mode 1)
(setq electric-pair-inhibit-predicate
      (lambda (char)
        (member major-mode '(org-mode))))
#+END_SRC

** Disable ESC ESC ESC

#+BEGIN_SRC emacs-lisp
(global-unset-key (kbd "ESC ESC"))
#+END_SRC

** Replace selected text
Enable =delete-selection-mode= so that if there is a marked region, typing in text replaces it

#+BEGIN_SRC emacs-lisp
(delete-selection-mode 1)
#+END_SRC

** Indentation: spaces, not tabs!
Set up the indentation behavior. I took this basic configuration from [[https://dougie.io/emacs/indentation/]]

#+begin_src emacs-lisp

;; Set the default tab settings
(setq-default tab-width 4)
(setq-default indent-tabs-mode nil)
(setq-default c-basic-offset 4)
(setq-default python-indent-offset 4)

;; Make the backspace properly erase the whole tab instead of removing
;; 1 space at a time
(setq backward-delete-char-untabify-method 'hungry)

#+end_src

** Files
Keybindings for working with files.

#+BEGIN_SRC emacs-lisp

  ;; Keep track of recently-opened files
  (use-package recentf
    :elpaca nil
    :config
    (recentf-mode 1)
    :custom
    (recentf-max-menu-items 5000)
    (recentf-max-saved-items 10000)
    :bind
    ("C-x C-r" . consult-recent-file))
    ;; :general
    ;; (ns/leader-def "fr" 'consult-recent-file))
#+END_SRC

#+BEGIN_SRC emacs-lisp :tangle nil
  (ns/leader-def
    "." 'find-file
    "f" '(:ignore t :which-key "files")
    "fs" 'find-file)
#+END_SRC

** Buffers
#+BEGIN_SRC emacs-lisp :tangle nil
(ns/leader-def
  "," '(consult-buffer :which-key "select buffer")
  "b" '(:ignore t :which-key "buffers")
  "bb" '(consult-buffer :which-key "select buffer"))
#+END_SRC

** Windows
Keybindings for operating windows.

#+BEGIN_SRC emacs-lisp
(use-package ace-window
  :bind ("M-o" . ace-window))
#+END_SRC

** Text manipulation

#+BEGIN_SRC emacs-lisp :tangle nil
(ns/leader-def
  "/" '(comment-line :which-key "Toggle comment"))
#+END_SRC

** Eval
#+BEGIN_SRC emacs-lisp :tangle nil
(ns/leader-def
  "e" '(:ignore t :which-key "eval")
  "el" '(eval-last-sexp :which-key "eval last sexpr")
  "ed" '(eval-defun :which-key "eval defun")
  "e:" '(eval-expression :which-key "eval expression"))
#+END_SRC

** Dumb-jump

#+BEGIN_SRC emacs-lisp
(use-package hydra)

(use-package dumb-jump
  :config
  (defhydra dumb-jump-hydra (:color blue :columns 3)
    "Dumb Jump"
    ("j" dumb-jump-go "Go")
    ("o" dumb-jump-go-other-window "Other window")
    ("e" dumb-jump-go-prefer-external "Go external")
    ("x" dumb-jump-go-prefer-external-other-window "Go external other window")
    ("i" dumb-jump-go-prompt "Prompt")
    ("l" dumb-jump-quick-look "Quick look")
    ("b" dumb-jump-back "Back"))
  (keymap-global-set "M-g j" 'dumb-jump-hydra/body))
#+END_SRC

** evil-textobj-tree-sitter
#+BEGIN_SRC emacs-lisp :tangle nil
(use-package evil-textobj-tree-sitter
  :config
  ; functions
  (define-key evil-outer-text-objects-map "f"
              (evil-textobj-tree-sitter-get-textobj "function.outer"))
  (define-key evil-inner-text-objects-map "f"
              (evil-textobj-tree-sitter-get-textobj "function.inner"))
  (define-key evil-normal-state-map
              (kbd "]f")
              (lambda ()
                (interactive)
                (evil-textobj-tree-sitter-goto-textobj "function.outer")))
  (define-key evil-normal-state-map
              (kbd "[f")
              (lambda ()
                (interactive)
                (evil-textobj-tree-sitter-goto-textobj "function.outer" t))))
#+END_SRC

* Configure builtin/core Emacs packages
** Misc
#+BEGIN_SRC emacs-lisp
(use-package emacs
  :elpaca nil
  :custom
  (scroll-margin 5)
  (ring-bell-function 'ignore)
  :config
  (pixel-scroll-mode 1)
  (global-visual-line-mode 1))
#+END_SRC

** Saveplace: Remember your location in a file
#+BEGIN_SRC emacs-lisp

(use-package saveplace
  :elpaca nil
  :unless noninteractive
  :config
  (setq save-place-limit 1000)
  (save-place-mode))

#+END_SRC 

** Savehist: Keep track of minibuffer history
#+BEGIN_SRC emacs-lisp

(use-package savehist
  :elpaca nil
  :unless noninteractive
  :defer 1
  :config
  (setq savehist-additional-variables '(compile-command kill-ring regexp-search-ring))
  (savehist-mode 1))

#+END_SRC 

** +Global-auto-revert-mode+
Auto-revert all buffers by default

#+BEGIN_SRC emacs-lisp :tangle nil
(global-auto-revert-mode 1)
#+END_SRC

** Backup and auto-save files
Put all backup and auto-save files in subfolders of the Emacs directory instead of littering the filesystem with them, including for Tramp sessions.
I got the basic code for this from [[https://emacs.stackexchange.com/questions/33/put-all-backups-into-one-backup-folder]]

#+BEGIN_SRC emacs-lisp
(let ((backup-dir (concat user-emacs-directory "backups"))
      (auto-saves-dir (concat user-emacs-directory "auto-saves")))
  (dolist (dir (list backup-dir auto-saves-dir))
    (when (not (file-directory-p dir))
      (make-directory dir t)))
  (setq backup-directory-alist `(("." . ,backup-dir))
        auto-save-file-name-transforms `((".*" ,auto-saves-dir t))
        tramp-backup-directory-alist `((".*" . ,backup-dir))
        tramp-auto-save-directory auto-saves-dir))

(setq backup-by-copying t   ; Don't delink hardlinks
      delete-old-versions t ; Clean up the backups
      version-control t     ; Use version numbers on backups
      kept-new-versions 2   ; Keep some new versions of backups
      kept-old-versions 1)  ; Keep some old backups too

(setq backup-directory-alist
      `(("." . ,(concat user-emacs-directory "backups/"))))
#+END_SRC

** Winner-mode: undo/redo window configuration changes
| Command     | Keybinding | Description                               |
|-------------+------------+-------------------------------------------|
| ~winner-undo~ | C-c left   | Undo the most recent window config change |
| ~winner-redo~ | C-c right  | Redo the most recent window config change |

#+BEGIN_SRC emacs-lisp
(winner-mode 1)
#+END_SRC

** display-line-numbers-mode: Displaying line numbers
#+BEGIN_SRC emacs-lisp
(use-package display-line-numbers-mode
  :elpaca nil
  :custom
  (display-line-numbers-type 'relative)
  :hook
  (prog-mode . display-line-numbers-mode)
  (markdown-mode . display-line-numbers-mode))
#+END_SRC

** Apply changes
Process the elpaca queue in case later packages depend on these settings
#+BEGIN_SRC emacs-lisp
(elpaca-wait)
#+END_SRC
* UI
Set up all the fancy user-interface elements to make Emacs look nice.

** Basic Emacs appearance settings
Set the window title to display the current file and major mode.

#+BEGIN_SRC emacs-lisp
(setq-default frame-title-format '("GNU Emacs"))
#+END_SRC 

** Diminish: keeping the modeline uncluttered
#+BEGIN_SRC emacs-lisp
(use-package diminish
  :init
  (diminish 'buffer-face-mode)
  (diminish 'visual-line-mode)
  (diminish 'abbrev))
#+END_SRC 

** Fonts
Set up the fonts for text rendering, pretty self explanatory :)

Here are my current font settings
#+BEGIN_SRC emacs-lisp
(defvar ns/default-font "JetBrainsMono NF"
  "My custom default font choice.")

(defvar ns/fixed-pitch-font "JetBrainsMono NF"
  "My custom fixed pitch font choice.")

(defvar ns/variable-pitch-font "JetBrainsMono NF"
  "My custom variable pitch font choice.")

(custom-set-faces
 `(default ((t (:family ,ns/default-font :slant normal :weight regular :height 110 :width normal :foundry "JB  "))))
 `(fixed-pitch ((t (:family ,ns/fixed-pitch-font :height 110))))
 `(variable-pitch ((t (:family ,ns/variable-pitch-font)))))
#+END_SRC  

Also install helpful icons useful for a ton of packages to add more visual detail. *NOTE* you'll have to run ~M-x all-the-icons-install-fonts~ in order to, well, install the fonts!

#+BEGIN_SRC emacs-lisp
(use-package all-the-icons)
;; (use-package nerd-icons)
#+END_SRC 

** Theme
Give Emacs a nice color scheme! Other themes that I like include:
- doom-palenight
- doom-shades-of-purple
- doom-flatwhite
- doom-tomorrow-day
- mindre-theme

#+BEGIN_SRC emacs-lisp
(use-package doom-themes
  :config
  (load-theme 'doom-one t))

(use-package ef-themes
  :config
  (setq ef-themes-headings ; read the manual's entry or the doc string
      '((0 . (variable-pitch light 1.9))
        (1 . (variable-pitch light 1.8))
        (2 . (variable-pitch regular 1.7))
        (3 . (variable-pitch regular 1.6))
        (4 . (variable-pitch regular 1.5))
        (5 . (variable-pitch 1.4)) ; absence of weight means `bold'
        (6 . (variable-pitch 1.3))
        (7 . (variable-pitch 1.2))
        (t . (variable-pitch 1.1))))
  (setq ef-themes-to-toggle '(ef-summer ef-cherie)
        ef-themes-mixed-fonts t
        ef-themes-variable-pitch-ui t))
  
  ;; (ef-themes-select 'ef-summer))

(use-package modus-themes)
  ;; :init
  ;; (setq modus-themes-mode-line '(moody)))
  ;; :config
  ;; (load-theme 'modus-vivendi t))

#+END_SRC

** Modeline
*** doom-modeline
#+BEGIN_SRC emacs-lisp
(use-package doom-modeline
  :init
  (setq doom-modeline-height 35
        doom-modeline-support-imenu t)
  (doom-modeline-mode 1))
#+END_SRC

** Dashboard
I like having a nice dashboard when Emacs loads in order to remember what I was last working on and (eventually!) view my ~org-agenda~ for the day. Maybe one day I'll revert to just a scratch buffer like others, but I'm sticking with this for now!

#+BEGIN_SRC emacs-lisp

;; Necessary for dashboard in order to get nice seperators between sections
(use-package page-break-lines)

(use-package dashboard
    :custom
    (dashboard-image-banner-max-width 256)
    (dashboard-startup-banner (expand-file-name "emacs.png" user-emacs-directory))
    ;; (dashboard-startup-banner "/Users/nsalesky/Pictures/wallpapers/forest-small.jpg")
    (dashboard-center-content t)
    (dashboard-set-heading-icons t)
    (dashboard-set-file-icons t)
    (dashboard-projects-backend 'project-el)
    ;; (dashboard-projects-switch-function 'projectile-persp-switch-project)
    (dashboard-items '((recents . 5)
                          (projects . 5)
                          (agenda . 5)))
    (initial-buffer-choice (lambda () (get-buffer-create "*dashboard*")))
    ;; :hook (after-init-hook . dashboard-refresh-buffer)
    :config
    (dashboard-setup-startup-hook))

#+END_SRC

** Helpful
Make the Emacs help pages more "helpful".

#+begin_src emacs-lisp
(use-package helpful
  :bind
  (("C-h f" . helpful-callable)
   ("C-h v" . helpful-variable)
   ("C-h k" . helpful-key)))

  ;; :general
  ;; (ns/leader-def
  ;;   "h" '(:ignore t :which-key "help")
  ;;   "hv" '(helpful-variable :which-key "describe variable")
  ;;   "hf" '(helpful-callable :which-key "describe function")
  ;;   "hk" '(helpful-key :which-key "describe key")))
#+end_src

** Hl-todo: Highlight and navigate TODO keywords

#+BEGIN_SRC emacs-lisp
(use-package hl-todo
  :config
  (global-hl-todo-mode))
#+END_SRC

** Transparent Frame
Creates a handy little function to toggle a semi-transparent window frame either for aeshetics or to view another window underneath Emacs.

#+BEGIN_SRC emacs-lisp
(defun ns/toggle-window-transparency ()
  "Toggle transparency."
  (interactive)
  (let ((alpha-transparency 90))
    (if (equal alpha-transparency (frame-parameter nil 'alpha-background))
        (set-frame-parameter nil 'alpha-background 100)
      (set-frame-parameter nil 'alpha-background alpha-transparency))))

;; Make the frame transparent when launched
;; (ns/toggle-window-transparency)
#+END_SRC

** Discover.el

#+BEGIN_SRC emacs-lisp
(use-package discover)
#+END_SRC

* Minibuffer
Sets up minibuffer completion with Vertico.

** Vertico
#+BEGIN_SRC emacs-lisp
(use-package vertico
  :init
  (vertico-mode))
#+END_SRC 

** Consult: add helpful functions with completion

#+BEGIN_SRC emacs-lisp
(use-package consult
  :bind
  (;; C-c bindings
   ("C-c h" . consult-history)
   ("C-c m" . consult-mode-command)
   ("C-c k" . consult-kmacro)
   ; M-g bindings
   ("M-g g" . consult-goto-line)
   ("M-g M-g" . consult-goto-line)
   ("M-g i" . consult-imenu)

   ; Buffers
   ("C-x b" . consult-buffer)
   ("C-x 4 b" . consult-buffer-otther-window)
   ("C-x p b" . consult-project-buffer)

   ; Random
   ("C-x r b" . consult-bookmark)
   ("M-y" . consult-yank-pop)

   ; Special search bindings
   ("C-c q" . consult-line)
   ("C-c w" . consult-ripgrep)
   
   ; M-s bindings (search-map)
   ("M-s d" . consult-find)
   ("M-s r" . consult-ripgrep)
   ("M-s l" . consult-line)
   ("M-s L" . consult-line-multi)
   ("M-s k" . consult-keep-lines)
   ("M-s u" . consult-focus-lines)

   ; Isearch integration
   ("M-s e" . consult-isearch-history)
   :map isearch-mode-map
   ("M-e" . consult-isearch-history)
   ("M-s e" . consult-isearch-history)
   ("M-s l" . consult-line)
   ("M-s L" . consult-line-multi)

   ; Minibuffer history
   :map minibuffer-local-map
   ("M-s" . consult-history)
   ("M-r" . consult-history))

  ;; :general
  ;; (ns/leader-def
  ;;   "s" '(:ignore t :which-key "search")
  ;;   "sr" '(consult-ripgrep :which-key "ripgrep")
  ;;   "sl" '(consult-line :which-key "line search")
  ;;   "si" '(consult-imenu :which-key "imenu"))

  :init
  (setq consult-narrow-key (kbd "<"))

  ;; Projectile
  ;; (autoload 'projectile-project-root "projectile")
  ;; (setq consult-project-function (lambda (_) (projectile-project-root)))

  ;; Project.el
  (setq consult-project-function #'consult--default-project-function))
#+END_SRC 

** Orderless: match the different parts of completion candidates

#+BEGIN_SRC emacs-lisp

(use-package orderless
  :custom
  (completion-styles '(orderless basic))
  (completion-category-overrides '(
                                   (file (styles basic partial-completion))
                                   (eglot (styles orderless)))))

#+END_SRC 

** Marginalia: decorate minibuffer completion candidates

#+BEGIN_SRC emacs-lisp
(use-package marginalia
  :bind
  (:map minibuffer-local-map
        ("M-A" . marginalia-cycle))
  :custom
  (marginalia-align 'right)
  :init
  (marginalia-mode)
  :config
  (setq marginalia-command-categories
        (append '((projectile-find-file . project-file)
                  (projectile-find-dir . project-file)
                  (projectile-switch-project . file))
                marginalia-command-categories)))
#+END_SRC

** All-the-icons-completion: add icons to completion candidates

#+BEGIN_SRC emacs-lisp

(use-package all-the-icons-completion
  :after (marginalia all-the-icons)
  :hook (marginalia-mode . all-the-icons-completion-marginalia-setup)
  :init
  (all-the-icons-completion-mode))

#+END_SRC 

** Embark

#+BEGIN_SRC emacs-lisp

(use-package embark
  ;; TODO: set up bindings for embark-act and embark-dwim
  :bind
  (("C-." . embark-act)
   ("M-." . embark-dwim))

  :config

  ;; Hide the mode line for Embark buffers
  (add-to-list 'display-buffer-alist
               '("\\`\\*Embark Collect \\(Live\\|Completions\\)\\*"
                 nil
                 (window-parameters (mode-line-format . none)))))

(use-package embark-consult
  :after (embark consult)
  :demand t
  :hook
  (embark-collect-mode . consult-preview-at-point-mode))

#+END_SRC 

* Editor

.doom.d/config.el
(setq user-full-name "Nick Salesky"
      user-mail-address "nicksalesky@gmail.com")

(setq doom-font (font-spec :family "JetBrainsMono" :size 16)
      doom-variable-pitch-font (font-spec :family "SourceSans3" :size 15)
      doom-big-font (font-spec :family "JetBrainsMono" :size 24))
(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))
(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))

(setq doom-theme 'doom-moonlight)
(map! :leader
      :desc "Load new theme" "h t" #'counsel-load-theme)

(set-face-attribute 'mode-line nil :font "JetBrainsMono")
(setq doom-modeline-height 30     ;; sets modeline height
      doom-modeline-bar-width 5   ;; sets right bar width
      doom-modeline-persp-name t  ;; adds perspective name to the modeline
      doom-modeline-persp-icon t) ;; adds a folder icon next to the perspective name

(setq frame-title-format
      '(""
        (:eval
         (if (s-contains-p org-roam-directory (or buffer-file-name ""))
             (replace-regexp-in-string
              ".*/[0-9]*-?" "☰ "
              (subst-char-in-string ?_ ? buffer-file-name))
           "%b"))
        (:eval
         (let ((project-name (projectile-project-name)))
           (unless (string= "-" project-name)
             (format (if (buffer-modified-p) " ◉ %s" "  ●  %s") project-name))))))

;; (if (eq initial-window-system 'x)
;;     (toggle-frame-maximized)
;;   (toggle-frame-fullscreen))

(defun ns/toggle-window-transparency ()
  "Toggle transparency."
  (interactive)
  (let ((alpha-transparency 75))
    (if (equal alpha-transparency (frame-parameter nil 'alpha-background))
        (set-frame-parameter nil 'alpha-background 100)
      (set-frame-parameter nil 'alpha-background alpha-transparency))))

(evil-global-set-key 'motion "j" 'evil-next-visual-line)
(evil-global-set-key 'motion "k" 'evil-previous-visual-line)

(map! :leader
      :desc "Org babel tangle" "m B" #'org-babel-tangle)
(after! org
  (setq org-directory "~/notes"
        org-agenda-files '("~/notes/agenda.org")
        org-default-notes-file (expand-file-name "notes.org" org-directory)
        org-return-follows-link t  ;; pressing RET follows links
        org-ellipsis " ▼ "
        org-superstar-headline-bullets-list '("◉" "●" "○" "◆" "●" "○" "◆")
        org-log-done 'time
        org-hide-emphasis-markers t))
        ;; org-todo-keywords ; This overwrites the default Doom org-todo-keywords
        ;; '((sequence
        ;;    "TODO(t)" ; A task that is ready to be tackled
        ;;    "WAIT(w)" ; Some thing that is holding up this task
        ;;    "HOLD(h)" ; Something that has been put on hold
        ;;    "|" ; Seperator between "active" and "inactive" states
        ;;    "DONE(d)" ; Task has been completed
        ;;    "CANCELLED(c)" ; Task has been canceled

(use-package org-modern
  :hook
  (org-mode . org-modern-mode)
  (org-agenda-finalize . org-modern-agenda))

;; (defun ns/org-mode-visual-fill ()
;;   (setq visual-fill-column-width 200
;;         visual-fill-column-center-text t)
;;   (visual-fill-column-mode 1))

;; (use-package visual-fill-column
;;   :hook (org-mode . ns/org-mode-visual-fill))

(custom-set-faces
  '(org-level-1 ((t (:inherit outline-1 :height 1.4))))
  '(org-level-2 ((t (:inherit outline-2 :height 1.3))))
  '(org-level-3 ((t (:inherit outline-3 :height 1.2))))
  '(org-level-4 ((t (:inherit outline-4 :height 1.1))))
  '(org-level-5 ((t (:inherit outline-5 :height 1.0))))
)

;(add-hook 'org-mode 'variable-pitch-mode)

(setq org-journal-dir "~/org/journal/"
      org-journal-date-prefix "* "
      org-journal-time-prefix "** "
      org-journal-date-format "%B %d, %Y (%A) "
      org-journal-file-format "%Y-%m-%d.org"
      org-journal-enable-agenda-integration t)

(after! org-roam
  (setq org-roam-directory "~/notes/roam"))

(after! org-roam
  (setq org-roam-capture-templates
       '(("d" ; the "key", a letter that you press to choose the template
          "default" ; the full name of the template
          plain ; the type of text being inserted, always =plain= for note templates
          "%?" ; the text that will be inserted into the new note, can be anything
          :if-new (file+head "%<%Y%m%d%H%M%S>-${slug}.org" "#+title: ${title}\n") ; this list describes how the note file will be created
          :unnarrowed t) ; ensures that the full file will be displayed when captured

         ;; A capture template for a programming language
         ("l" "programming language" plain
          (file "~/notes/roam/templates/programming-language-template.org")
          :if-new (file+head "%<%Y%m%d%H%M%S>-${slug}.org" "#+title: ${title}\n")
          :unnarrowed t)

         ;; A capture template for a project I'm working on
         ("p" "project" plain
          (file "~/notes/roam/templates/project-template.org")
          :if-new (file+head "%<%Y%m%d%H%M%S>-${slug}.org" "#+title: ${title}\n#+filetags: Project")
          :unnarrowed t)
                        )))

(use-package org-remark
  :init
  (map! :leader
        (:prefix ("r" . "remark")
         ;; custom pen bindings
         :desc "Mark yellow" "y" #'org-remark-mark-yellow
         :desc "Mark green" "g" #'org-remark-mark-green
         :desc "Mark red" "r" #'org-remark-mark-red

         ;; general bindings
         :desc "Open current remark" "o" #'org-remark-open
         :desc "View current remark" "v" #'org-remark-view
         :desc "View previous remark" "p" #'org-remark-view-prev
         :desc "View next remark" "n" #'org-remark-view-next
         :desc "Delete remark" "d" #'org-remark-remove))
  :config

  ;; set up my pens
  (org-remark-create "orange"
                     '(:underline "gold" :background "dark orange")
                     '(CATEGORY "important"))
  (org-remark-create "green"
                     '(:background "lime green" :inherit shadow)
                     '(CATEGORY "vocab"))
  (org-remark-create "red"
                     '(:background "tomato" :underline "dark red" :inherit shadow)
                     '(CATEGORY "important"))

  (org-remark-global-tracking-mode))

(setq display-line-numbers-type t)
(map! :leader
      :desc "Comment or uncomment lines" "TAB TAB" #'comment-line
      (:prefix ("t" . "toggle")
       :desc "Toggle line numbers" "l" #'doom/toggle-line-numbers
       :desc "Toggle line highlight in frame" "h" #'hl-line-mode
       :desc "Toggle line highlight globally" "H" #'global-hl-line-mode
       :desc "Toggle truncate lines" "t" #'toggle-truncate-lines))

(setq confirm-kill-emacs nil)

(setq auto-save-default t
      auto-save-timeout 30)

(map! :leader
      (:prefix ("e". "evaluate")
       :desc "Evaluate elisp in buffer" "b" #'eval-buffer
       :desc "Evaluate defun" "d" #'eval-defun
       :desc "Evaluate elisp expression" "e" #'eval-expression
       :desc "Evaluate last sexpr" "l" #'eval-last-sexp
       :desc "Evaluate elisp in region" "r" #'eval-region))

(map! :leader
      (:prefix ("d" . "dired")
       :desc "Open dired" "d" #'dired
       :desc "Dired jump to current" "j" #'dired-jump)
      (:after dired
       (:map dired-mode-map
        :desc "Peep-dired image previews" "d p" #'peep-dired
        :desc "Dired view file" "d v" #'dired-view-file)))

(evil-define-key 'normal dired-mode-map
  (kbd "M-RET") 'dired-display-file
  (kbd "h") 'dired-up-directory
  (kbd "l") 'dired-open-file ; use dired-find-file instead of dired-open.
  (kbd "m") 'dired-mark
  (kbd "t") 'dired-toggle-marks
  (kbd "u") 'dired-unmark
  (kbd "C") 'dired-do-copy
  (kbd "D") 'dired-do-delete
  (kbd "J") 'dired-goto-file
  (kbd "M") 'dired-do-chmod
  (kbd "O") 'dired-do-chown
  (kbd "P") 'dired-do-print
  (kbd "R") 'dired-do-rename
  (kbd "T") 'dired-do-touch
  (kbd "Y") 'dired-copy-filenamecopy-filename-as-kill ; copies filename to kill ring.
  (kbd "+") 'dired-create-directory
  (kbd "-") 'dired-up-directory
  (kbd "% l") 'dired-downcase
  (kbd "% u") 'dired-upcase
  (kbd "; d") 'epa-dired-do-decrypt
  (kbd "; e") 'epa-dired-do-encrypt)

;; Get file icons in dired
(add-hook 'dired-mode-hook 'all-the-icons-dired-mode)

;; With dired-open plugin you can launch external programs for certain extensions
;;(setq dired-open-extensions '(("jpg" . "gimp")
;;                              ("mp4" . "vlc" )))

(evil-define-key 'normal peep-dired-mode-map
  (kbd "j") 'peep-dired-next-file
  (kbd "k") 'peep-dired-prev-file)
(add-hook 'peep-dired-hook 'evil-normalize-keymaps)

;; (map! :leader
;;       :desc "Switch to perspective NAME" "DEL" #'persp-switch
;;       :desc "Switch to buffer in perspective" "," #'persp-switch-to-buffer
;;       :desc "Switch to next perspective" "]" #'persp-next
;;       :desc "Switch to previous perspective" "[" #'persp-prev
;;       :desc "Add a buffer current perspective" "+" #'persp-add-buffer
;;       :desc "Remove perspective by name" "-" #'persp-remove-by-name)

(use-package wakatime-mode
  :ensure t
  :config
  (global-wakatime-mode))
